### PR TITLE
[PasswordHasher] Add union types

### PR DESCRIPTION
**Pull request by ValentineBoineau**
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| New feature?  | no 
| Deprecations? | no 
| Tickets       | 
| License       | MIT
| Doc PR        | 

Extracted from #41424

### CODE REVIEW

File: src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactory.php [hunk 1]
```diff
@@ -38,7 +40,7 @@ public function __construct(array $passwordHashers)
     /**
      * {@inheritdoc}
      */
-    public function getPasswordHasher($user): PasswordHasherInterface
+    public function getPasswordHasher(UserInterface|string|PasswordAuthenticatedUserInterface $user): PasswordHasherInterface
```
Comment (stof) [inline src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactory.php]: This misses `PasswordHasherAwareInterface` (unless we expect to support that only in combination with `PasswordAuthenticatedUserInterface`)

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 2]
```diff
@@ -31,10 +31,7 @@ public function __construct(PasswordHasherFactoryInterface $hasherFactory)
         $this->hasherFactory = $hasherFactory;
     }
 
-    /**
-     * @param PasswordAuthenticatedUserInterface $user
-     */
-    public function hashPassword($user, string $plainPassword): string
+    public function hashPassword(UserInterface|string|PasswordAuthenticatedUserInterface $user, string $plainPassword): string
```
Comment (stof) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: strings are not supported in UserPasswordHasher.

And anything else than `PasswordAuthenticatedUserInterface` triggers a deprecation (and so will be removed before the 6.0 release), so the typehint should be `PasswordAuthenticatedUserInterface`, not `PasswordAuthenticatedUserInterface|UserInterface`

Comment (stof) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: same in other methods

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php (hunk 2 above)
Comment (ValentineBoineau) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: I tried but the function will need some rework if we put only  `PasswordAuthenticatedUserInterface` , there is an error for `$user->getSalt` but I don't know how to handle it

Comment (stof) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: Well, that will be solved when cleaning deprecated APIs. In 6.0, `getSalt` should be called only if the User implements `LegacyPasswordAuthenticatedUserInterface`.

The issue here is that you try to implement the new typehints _before_ cleaning the deprecated code, and so you still need to account for old types while 6.0 won't need that.

Comment (ValentineBoineau) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: But what appends if the user is not `LegacyPasswordAuthenticatedUserInterface` ?

Comment (stof) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: `$salt = null` in Symfony 6

Comment (ValentineBoineau) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: Ok thank you!

File: src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactory.php (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactory.php]: `UserInterface` can be removed also.
```suggestion
    public function getPasswordHasher(string|PasswordAuthenticatedUserInterface|PasswordHasherAwareInterface $user): PasswordHasherInterface
```

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 3]
```diff
@@ -79,22 +59,12 @@ public function isPasswordValid($user, string $plainPassword): bool
...
             return false;
         }
 
-        if (!$user instanceof PasswordAuthenticatedUserInterface) {
-            if (!$user instanceof UserInterface) {
-                throw new \TypeError(sprintf('Expected an instance of "%s" as first argument, but got "%s".', UserInterface::class, get_debug_type($user)));
-            }
-            trigger_deprecation('symfony/password-hasher', '5.3', 'The "%s()" method expects a "%s" instance as first argument. Not implementing it in class "%s" is deprecated.', __METHOD__, PasswordAuthenticatedUserInterface::class, get_debug_type($user));
-        }
-
```
Comment (chalasr) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:97]: There is no union type on this method, this goes outside the scope of this PR :)

File: src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactoryInterface.php [hunk 4]
```diff
@@ -26,9 +25,9 @@ interface PasswordHasherFactoryInterface
     /**
      * Returns the password hasher to use for the given user.
      *
-     * @param PasswordAuthenticatedUserInterface|UserInterface|string $user A PasswordAuthenticatedUserInterface/UserInterface instance or a class name
+     * @param PasswordHasherAwareInterface|PasswordAuthenticatedUserInterface|string $user A PasswordAuthenticatedUserInterface/UserInterface instance or a class name
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactoryInterface.php]: let's remove the annotation altogether

File: src/Symfony/Component/PasswordHasher/Tests/Hasher/PasswordHasherFactoryTest.php [hunk 5]
```diff
@@ -21,18 +21,19 @@
 use Symfony\Component\PasswordHasher\PasswordHasherInterface;
 use Symfony\Component\Security\Core\Encoder\PlaintextPasswordEncoder;
 use Symfony\Component\Security\Core\User\InMemoryUser;
+use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;
 use Symfony\Component\Security\Core\User\UserInterface;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PasswordHasher/Tests/Hasher/PasswordHasherFactoryTest.php]: should be removed now

**Review by nicolas-grekas — APPROVED** [review]
With some nitpicking.
@chalasr shouldn't we backport some changes to 5.3?

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 6]
```diff
@@ -31,43 +30,23 @@ public function __construct(PasswordHasherFactoryInterface $hasherFactory)
         $this->hasherFactory = $hasherFactory;
     }
 
-    /**
-     * @param PasswordAuthenticatedUserInterface $user
-     */
-    public function hashPassword($user, string $plainPassword): string
+    public function hashPassword(PasswordAuthenticatedUserInterface $user, string $plainPassword): string
```
Comment (chalasr) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:33]:
```suggestion
    public function hashPassword(PasswordAuthenticatedUserInterface|PasswordHasherAwareInterface $user, string $plainPassword): string
```

(+ use statement)

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 7]
```diff
@@ -31,43 +30,23 @@ public function __construct(PasswordHasherFactoryInterface $hasherFactory)
...
         $hasher = $this->hasherFactory->getPasswordHasher($user);
 
         return $hasher->hash($plainPassword, $salt);
     }
 
-    /**
-     * @param PasswordAuthenticatedUserInterface $user
-     */
-    public function isPasswordValid($user, string $plainPassword): bool
+    public function isPasswordValid(PasswordAuthenticatedUserInterface $user, string $plainPassword): bool
```
Comment (chalasr) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:45]: same

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 8]
```diff
@@ -79,22 +58,12 @@ public function isPasswordValid($user, string $plainPassword): bool
         return $hasher->verify($user->getPassword(), $plainPassword, $salt);
     }
 
-    /**
-     * @param PasswordAuthenticatedUserInterface $user
-     */
-    public function needsRehash($user): bool
+    public function needsRehash(PasswordAuthenticatedUserInterface $user): bool
```
Comment (chalasr) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:61]: same

**Review by chalasr — CHANGES_REQUESTED** [review]
Rebase needed after #41678

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php (hunk 6 above)
Comment (stof) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:33]: Do we actually want to support the case of implementing PasswordHasherAwareInterface on a class that is not a PasswordAuthenticatedUserInterface ?

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php (hunk 7 above)
Comment (stof) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:45]: this method definitely requires a `PasswordAuthenticatedUserInterface` to be able to use `getPassword`

File: src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactoryInterface.php [hunk 9]
```diff
@@ -26,9 +25,7 @@ interface PasswordHasherFactoryInterface
     /**
      * Returns the password hasher to use for the given user.
      *
-     * @param PasswordAuthenticatedUserInterface|UserInterface|string $user A PasswordAuthenticatedUserInterface/UserInterface instance or a class name
```
Comment (stof) [inline src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactoryInterface.php]: the description was still explaining what type of string was supported

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php (hunk 6 above)
Comment (chalasr) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:33]: Indeed, it does not make sense here: `UserPasswordHasher` calls `PasswordAuthenticatedUserInterface#getPassword()` internally. I'm going to fix that on 5.3.
However allowing both types independently works in PasswordHasherFactory, should we keep it? I tend to think it makes sense.

Comment (chalasr) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:33]: see #41682

File: src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactoryInterface.php (hunk 9 above)
Comment (chalasr) [reply src/Symfony/Component/PasswordHasher/Hasher/PasswordHasherFactoryInterface.php]: any string is supported actually (named hashers)

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @ValentineBoineau.

**Pull request merged**